File: em-april-2026/dsa/leetcode_practice_problems/unsorted/02390_Removing_stars_from_a_string/main.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <stack>
#include <string>
#include <vector>

using namespace std;

class Solution {
   public:
    string removeStars(const string& s) {
        stack<int> st;
        for (auto c : s) {
            if (c == '*') {
                if (st.empty()) {
                    throw "stack is empty. input is not proper.";
                } else {
                    st.pop();
                }
            } else {
                st.push(c);
            }
        }
        auto ret = vector<char>();
        while (!st.empty()) {
            ret.emplace_back(st.top());
            st.pop();
        }
        reverse(ret.begin(), ret.end());
        return string(ret.cbegin(), ret.cend());
    }
};

bool Test(string s, string expected) {
    auto orig_s = s;
    cout << "[RUN]\n  s=\'" << s << "\', expected_answer=\'" << expected << "\'"
         << endl;
    auto actual = Solution().removeStars(s);
    if (actual != expected) {
        cerr << "  actual=\'" << actual << "\', expected=\'" << expected
             << "\'\n[FAILED]" << endl;
        return false;
    }
    cout << "[DONE]" << endl;
    return true;
}

int main() {
    if (!Test("leet**cod*e", "lecoe")) exit(1);
    if (!Test("erase*****", "")) exit(1);
    if (!Test("abc**", "a")) exit(1);
    if (!Test("ab**", "")) exit(1);
    if (!Test("a*", "")) exit(1);
    if (!Test("", "")) exit(1);
    return 0;
}

File: em-april-2026/dsa/leetcode_practice_problems/unsorted/02390_Removing_stars_from_a_string/main.py
class Solution:
    def removeStars(self, s: str) -> str:
        # Simple O(n), (s) solution
        stack = []
        for c in s:
            if c == "*":
                if not stack:
                    raise ValueError(
                        f"stack is empty. Input string '{s}' is not proper."
                    )
                stack.pop()
            else:
                stack.append(c)
        return "".join(stack)


def Test(s: str, expected: str):
    print(f"[RUN]\n  s='{s}', expected_answer='{expected}'")
    actual = Solution().removeStars(s)
    assert actual == expected, f"  actual='{actual}', expected='{expected}'\n[FAILED]"
    print("[DONE]")


Test("leet**cod*e", "lecoe")
Test("erase*****", "")
Test("abc**", "a")
Test("ab**", "")
Test("a*", "")
Test("", "")
